File: cwe_comparative/CWE-476_null_dereference/vulnerable.go
// CWE-476: NULL Pointer Dereference (Go: nil pointer panic)
package main

import "fmt"

type User struct {
	ID   int
	Name string
}

func findUser(id int) *User {
	if id == 42 {
		return &User{ID: 42, Name: "alice"}
	}
	return nil // user not found
}

func main() {
	u := findUser(99)      // not found → nil
	fmt.Println(u.Name)   // PANIC: nil pointer dereference
}

File: cwe_comparative/CWE-476_null_dereference/vulnerable.rs
// CWE-476: NULL Pointer Dereference (Rust: .unwrap() on None panics)
struct User {
    id: u32,
    name: String,
}

fn find_user(id: u32) -> Option<User> {
    if id == 42 {
        Some(User { id: 42, name: "alice".to_string() })
    } else {
        None
    }
}

fn main() {
    // PANIC: called `Option::unwrap()` on a `None` value
    let u = find_user(99).unwrap();
    println!("User: {}", u.name);
}
